LanguageFeatures/Static-access-shorthand/constant_expression_A03_t03.dart

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Given an expression that is a prefix of
+/// `<staticMemberShorthandHead> <selector>*`, whose assigned shorthand context
+/// denotes a declaration D, and where the identifier or `new` of the
+/// `<staticMemberShorthandHead>` denotes a static declaration or constructor
+/// declaration S when looked up on D.
+/// ...
+/// - An expression of the form `.<identifier>` that is not followed by an
+///   `<argumentPart>`, is a constant expression if S declares a static method
+///   or constructor with base name `<identifier>`, and either type inference
+///   has not added type arguments as a generic function instantiation coercion
+///   to the method, or to the target class for a constructor, or the added type
+///   arguments are constant types.
+///
+/// @description Checks that an expression of the form `'.'<identifier>` that is
+/// not followed by an `<argumentPart>` is not a constant expression if the
+/// type argument is not a constant expression.
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=enum-shorthands
+
+import '../../Utils/expect.dart';
+
+class C {
+  static String foo<X>() => "C<$X>";
+
+  @override
+  bool operator ==(Object other) {
+    if (other is Function) {
+      Expect.equals("C<String>", other());
+      Expect.equals(C.foo<String>, other);
+      return identical(C.foo<String>, other);
+    }
+    return true;
+  }
+}
+
+mixin M {
+  static String foo<X>() => "M<$X>";
+
+  @override
+  bool operator ==(Object other) {
+    if (other is Function) {
+      Expect.equals("M<String>", other());
+      Expect.equals(M.foo<String>, other);
+      return identical(M.foo<String>, other);
+    }
+    return true;
+  }
+}
+class MA = Object with M;
+
+test<T>() {
+  Expect.isFalse(C() == .foo<T>);
+  Expect.isFalse((MA() as M) == .foo<T>);
+}
```

Comment:
I'm a little bit worried about this because it _requires_ certain torn-off functions to be non-canonicalized. I don't think we want to prevent it if a compiler will actually detect that it can apply this optimization in some cases.

We could test that it doesn't crash, and ignore the returned boolean value.

I think it would be nice to have a comment, as well, explaining why we do not require the result to be false.

Reply:
Thanks! I really want to avoid preventing those optimizations, so it's a good thing to keep an eye on things that would outlaw them.